crates/apollo-compiler/src/validation/input_object.rs

```diff
@@ -2,9 +2,72 @@ use std::{collections::HashMap, sync::Arc};
 
 use crate::{
     diagnostics::{ApolloDiagnostic, DiagnosticData, Label},
-    hir, ValidationDatabase,
+    hir,
+    validation::RecursionStack,
+    ValidationDatabase,
 };
 
+// Implements [Circular References](https://spec.graphql.org/October2021/#sec-Input-Objects.Circular-References)
+// part of the input object validation spec.
+struct FindRecursiveInputValue<'a> {
+    db: &'a dyn ValidationDatabase,
+}
+
+impl FindRecursiveInputValue<'_> {
+    fn input_value_definition(
+        &self,
+        seen: &mut RecursionStack<'_>,
+        def: &hir::InputValueDefinition,
+    ) -> Result<(), hir::InputValueDefinition> {
+        let ty = def.ty();
+        return match ty {
+            hir::Type::NonNull { ty, loc: _ } => match ty.as_ref() {
+                // NonNull type followed by Named type is the one that's not allowed
+                // to be cyclical, so this is only case we care about.
+                //
+                // Everything else may be a cyclical input value.
+                hir::Type::Named { name, loc: _ } => {
+                    if seen.contains(name) {
```

Comment:
In the recursive directive definition code I also checked `seen.first()` here because otherwise, an input type `B` referencing an input type `A` where `A` is cyclical would report an error on the definition of `B`. For example:

```graphql
input A {
  cycle: A!
}
input B {
  field: A!
}
```
![image](https://user-images.githubusercontent.com/1006268/228849716-219b9084-8b1d-46e0-83cc-e91be9de74f2.png)

Here the error you get on `B` is quite confusing, and unnecessary as you already have the error on `A`, and that one shows what you need to solve.

So I think we should do the same here (just return `Ok(())` if there is a cycle in a different input type than we started with)

Reply:
but what happens if you have two inputs that reference each other? if you only check the first item in the stack, you end up with something like this being valid:

```graphql
input A {
  cycle: B!
}

input B {
  cycle: A!
}
```

Reply:
does that ever happen with directives? ^